compiler/types.nim

```diff
@@ -587,79 +606,106 @@ proc typeToString(typ: PType, prefer: TPreferedDesc = preferName): string =
+        of tySet, tyVar: result = firstValue[T](t.sons[0])
+        of tyArray: result = firstValue[T](t.sons[0])
+        of tyInt:
+          if platform.intSize == 4: result = - (2147483646) - 2
+          else: result = 0x8000000000000000'i64
+        of tyInt8: result = - 128
+        of tyInt16: result = - 32768
+        of tyInt32: result = - 2147483646 - 2
+        of tyInt64: result = 0x8000000000000000'i64
+        of tyUInt..tyUInt64: result = 0
+        of tyEnum:
+          # if basetype <> nil then return firstOrd of basetype
+          if sonsLen(t) > 0 and t.sons[0] != nil:
+            result = firstValue[T](t.sons[0])
+          else:
+            assert(t.n.sons[0].kind == nkSym)
+            result = t.n.sons[0].sym.position
+        of tyOrdinal:
+          if t.len > 0: result = firstValue[T](lastSon(t))
+          else: internalError("invalid kind for first(" & $t.kind & ')')
+        else:
+          internalError("invalid kind for first(" & $t.kind & ')')
+          result = 0
+
+      elif T is BiggestFloat:
+        case t.kind:
+          of tyFloat..tyFloat128: result = -Inf
+          else:
+            internalError("invalid kind for first(" & $t.kind & ')')
+            result = NaN
+
+proc firstOrd*(t: PType): BiggestInt {.inline.} =
+  firstValue[BiggestInt](t)
+
+proc lastValue*[T:BiggestInt|BiggestFloat](t: PType; fixedUnsigned = false): T =
   case t.kind
-  of tyBool: result = 1
-  of tyChar: result = 255
-  of tySet, tyVar: result = lastOrd(t.sons[0])
-  of tyArray: result = lastOrd(t.sons[0])
+  of tyVar: result = lastValue[T](t.sons[0])
+  of tyGenericInst, tyDistinct, tyTypeDesc, tyAlias:
+    result = lastValue[T](lastSon(t))
   of tyRange:
     assert(t.n != nil)        # range directly given:
     assert(t.n.kind == nkRange)
-    result = getOrdValue(t.n.sons[1])
-  of tyInt:
-    if platform.intSize == 4: result = 0x7FFFFFFF
-    else: result = 0x7FFFFFFFFFFFFFFF'i64
-  of tyInt8: result = 0x0000007F
-  of tyInt16: result = 0x00007FFF
-  of tyInt32: result = 0x7FFFFFFF
-  of tyInt64: result = 0x7FFFFFFFFFFFFFFF'i64
-  of tyUInt:
-    if platform.intSize == 4: result = 0xFFFFFFFF
-    elif fixedUnsigned: result = 0xFFFFFFFFFFFFFFFF'i64
-    else: result = 0x7FFFFFFFFFFFFFFF'i64
-  of tyUInt8: result = 0xFF
-  of tyUInt16: result = 0xFFFF
-  of tyUInt32: result = 0xFFFFFFFF
-  of tyUInt64:
-    if fixedUnsigned: result = 0xFFFFFFFFFFFFFFFF'i64
-    else: result = 0x7FFFFFFFFFFFFFFF'i64
-  of tyEnum:
-    assert(t.n.sons[sonsLen(t.n) - 1].kind == nkSym)
-    result = t.n.sons[sonsLen(t.n) - 1].sym.position
-  of tyGenericInst, tyDistinct, tyTypeDesc, tyAlias:
-    result = lastOrd(lastSon(t))
-  of tyProxy: result = 0
-  of tyOrdinal:
-    if t.len > 0: result = lastOrd(lastSon(t))
-    else: internalError("invalid kind for last(" & $t.kind & ')')
+    result = getValue[T](t.n.sons[1])
   else:
-    internalError("invalid kind for last(" & $t.kind & ')')
-    result = 0
+    when T is BiggestInt:
```

Comment:
I dislike this style. It's not a "generic", it's 2-3 different implementations which happen to use the same name. And they are not generic.

Reply:
done, Introduced separate firstFloat and lastFloat procs